data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,675 @@
+
+In computer science language, the type arrow operator _associates to the right_. Function types without parentheses should be treated as if they have parentheses to the right in the same way that the type of `dummy_cat` was declared above. Except they are not displayed. This is [pretty printing](https://en.wikipedia.org/wiki/Prettyprint).
+
+### Passing Tuples
+
+In OCaml, a *tuple* is a data structure used to group a fixed number of values, which can be of different types. Tuples are surrounded by parentheses, and the elements are separated by commas. Here's the basic syntax to create and work with tuples in OCaml:
+```ocaml
+# ("felix", 1920);;
+- : string * int = ("felix", 1920)
+```
+
+It is possible to use the tuple syntax to specify function parameters. Here is how it can be used to define yet another version of the running example:
+```ocaml
+# let spicy_cat (x, y) = x ^ " " ^ y;;
+val spicy_cat : string * string -> string = <fun>
+```
+
+It behaves the same as previously.
+```ocaml
+# spicy_cat ("hello", "world");;
+- : string = "hello world"
+```
+
+It looks like two arguments have been passed: `"hello"` and `"world"`. However, only one, the `("hello", "world")` pair, has been passed. Inspection of the generated assembly would show it isn't exactly the same function. It contains some more code. The contents of the pair passed to `spicy_cat` (`x` and `y`) must be extracted before evaluation of the `s ^ " " ^ t` expression. This is the role of the additional assembly instructions.
+
+In many imperative languages, the `spicy_cat ("hello", "world")` syntax reads as a function call with two parameters; but in OCaml, it denotes applying the function `spicy_cat` to a pair containing `"hello"` and `"world"`.
+
+### Currying and Uncurrying
+
+In the previous sections, two kinds of multiple-parameter functions have been presented.
+- Functions returning a function, such as `sweet_cat` and `sour_cat`
+- Functions taking a tuple as a parameter, such as `spicy_cat`
+
+Interestingly, both kinds of functions provide a way to pass several pieces of data to a body while being functions with a single parameter. From this perspective, it makes sense to say: “All functions have a single argument.”
+
+This goes even further. It is always possible to translate back and forth between functions that look like `sweet_cat` (or `sour_cat`) and functions that look like `spicy_cat`.
+
+These translations have names:
+* [Currying](https://en.wikipedia.org/wiki/Currying) goes from the `spicy_cat` form into the `sour_cat` (or `sweet_cat`) form.
+* Uncurrying goes from the `sour_cat` (or `sweet_cat`) form into the `spicy_cat` form.
+
+It also said that `sweet_cat` and `sour_cat` are _curried_ functions whilst `spicy_cat` is _uncurried_.
+
+These translations are attributed to the 20th-century logician [Haskell Curry](https://en.wikipedia.org/wiki/Haskell_Curry).
```

Comment:
```suggestion
These definitions are attributed to the 20th-century logician [Haskell Curry](https://en.wikipedia.org/wiki/Haskell_Curry).
```

Do you like this word better, as it's not really a translation (or is it?)

Reply:
I've used _translation_ here to refer to what's been described above (and called translation). It is not a translation in the sense of translation between French and English. It is like turning a sentence from active to passive (or back)